Initial version of the library completed

zah opened the pull request.

kayabaNerve committed:
Standardize project structure around json_serialization.

kayabaNerve committed:
Split the monolithic protobuf_serialization into types/reader/writer.

kayabaNerve committed:
Fix usage of inline.

kayabaNerve committed:
Remove all function prototypes/lack of generics from reader.

Reader was done first to verify it didn't break compatibility with 
Writer. Now that the functionality has been confirmed to be the same, 
Writer will be redone.

From there, I plan to rework the test to include protobuf vectors and 
start creating types to match json-serialization.

kayabaNerve committed:
Use enumSerializedFields.

kayabaNerve committed:
Redo Reader.

This started as a move to streams. Then, when implementing the functions 
to read from a string, after attempting a lookup table, a new case 
statement was written. Then types were overhauled.

This ends up completely redoing the Reader. Support is preserved for 
VarInt and strings (user-defined length delimited types are currenty 
untested). Zigzags ints are broken with them being the next target.

Once the Reader is re-implemented, the API of test_serialization will be 
updated. If it passed, meaning the Reader is valid, the Writer will be 
re-implemented.

kayabaNerve committed:
Fix zigzag decoding.

kayabaNerve committed:
Implement Fixed32/Fixed64.

Still requires proper subtype detection.

kayabaNerve committed:
Expand uint to be either UInt32 or Fixed32.

This actually shouldn't be necessary here; only for the Writer, as this 
either/or is handled by the wire-type.

kayabaNerve committed:
Merge bit distinctions for sint/pint. Remove subtypes where possible.


This is really a cleanup commit.

Since we switch off the type, and error when the wrong pragma (sint32 
for int64) was attached, it's easy enough to make it so you can't attach 
the wrong pragma. Now it's just pint/sint.

Subtypes are only needed for VarInt. The only other ambiguous type is 
caused by an int/uint being able in a fixed format, yet due to the 
different wire type, we don't need to determine a subtype for it.

kayabaNerve committed:
Create a single template to get signed/unsigned integers.

kayabaNerve committed:
Update the API.

Not only does it now conform with the expected nim-serialization API, it 
updates to compile with faststreams. I believe I'm having version 
conflicts though, so these changes may need removal in the future.

kayabaNerve committed:
Rework the Reader to be as compile-time as possible.

Stops the need to have pointless stub functions for every type for every 
wire type.

kayabaNerve committed:
Update the writer to compile.

kayabaNerve committed:
Update the test API; comment out anything that requires length delimitation.

As of right now, there's a symbol resolution bug with it.

kayabaNerve committed:
Add test changes meant for the last commit.

kayabaNerve committed:
Stop the test from raising https://github.com/kayabaNerve/nim-protobuf-serialization/issues/1.

kayabaNerve committed:
Fix float encoding/decoding.

kayabaNerve committed:
Get length delimited fields working.

kayabaNerve committed:
Fix object decoding.

kayabaNerve committed:
Remove any guessing/defaulting to what type to use.

We only defaulted when handed a signed integer outside of an object, 
with no context on how to handle it. Even though we defaulted, this 
option was overridable.

Now, every signed integer requires explicit specification as to how to 
decode it. UInts and floats don't need this explicit specification due 
to a lack of ambiguity.

kayabaNerve committed:
Rework the typing structure.

To match the pint/sint/fixed/sfixed pragmas used for marking how a field 
in an object is encoded, PIntWrapped, SIntWrapped, UIntWrapped, 
FixedWrapped, and SFixedWrapped has been added for primitives outside of 
objects.

In order to keep with the requirement of not needing to specify bits, 
templates have been added to convert X to Y, as well as an X value to a 
Y value.

VarIntSubType was moved to Reader. The pragmas are enough for Writer. 
Reader also defines a new Protobuf error and removes all raises of the 
raw ProtobufError.

Also removes IntegerTypes, which was just being aliased to VarIntTypes 
and used as that, and removes RecognizedTypes. RecognizedTypes was just 
used to detect what's implicitly length delimited. The definition of 
LengthDelimitedTypes has been updated to compensate. It should be noted 
object parsing has broken because of that, yet due to the scope of these 
changes and the fact this isn't being used anywhere, I want to get this 
commit committed now.

Finally, legacy types have been explicitly marked as legacy.

kayabaNerve committed:
Fix a broken type definition preventing objects with chars from compiling.

kayabaNerve committed:
Enable nested object parsing.

kayabaNerve committed:
Finish the Reader rewrite.

It's not completely finished, as it'd benefit from increased checks. As 
of right now, I believe it's possible to provide either a VarInt or a 
Fixed for an uint field. While this won't decode the wrong value, it 
means two different objects will be treated with structural equality.

Enables nested object reading from a generated code stand point.

Fixes field number detection checking.

Adds a pragma for UInt fields (previously there was only Fixed, a 
variation of UInt).

Fixes 
https://github.com/kayabaNerve/nim-protobuf-serialization/issues/2.

Removes a test case. It encoded X, and then wrote X as a length 
delimited field, then expected the Reader to ignore the outer object for 
X alone.

kayabaNerve committed:
Add back the wireType template.

kayabaNerve committed:
Update legacy type definitions to new ones.

Helps frame the Writer's code. Some types weren't directly convertable 
due to how they've been replaced. SomeLengthDelimited has been 
transformed into CastableLengthDelimitedTypes, as well as 
LengthDelimitedTypes. The first is closest to the original, detailing 
anything castable from a seq[byte]. The second is actually any type 
which doesn't fit into any other type.

kayabaNerve committed:
Separate internal types/common data from ones meant to be exported.

kayabaNerve committed:
Rewrite the Writer.

kayabaNerve committed:
Add func/let/raises.

kayabaNerve committed:
Bug fixes.

Removes an infinite loop.

Fixes subtype detection for UIntWrapped types.

Fixes a template bug when converting types to wrapped types.

Partially fixes 
(https://github.com/kayabaNerve/nim-protobuf-serialization/issues/1). 
Now, non-objects won't keep reading from the buffer.

Partially edits type definitions. Insignificant for users, significant 
for the test I'm trying to get to pass.

kayabaNerve committed:
Port the libp2p VarInt test.

It doesn't test the libp2p modified VarInt variant, and it does 
partially fail tests it should pass. I believe that's due to the test 
being written for a different system though.

This test frequently violates type safety, casting uints to ints which 
are supposed to be negative and more. I understand the test was written 
for a different system, and makes sense in the context of this system, 
but porting it is still a WIP.

kayabaNerve committed:
Reorg the test directory.

kayabaNerve committed:
Mino style edits.

kayabaNerve committed:
Don't allow objects with more than 32 serialized fields.

Part of 
https://github.com/kayabaNerve/nim-protobuf-serialization/issues/5.

kayabaNerve committed:
Fix and test boolean handling.

Also counts seq[bool] as an CastableLengthDelimited type.

kayabaNerve committed:
Fix warnings/add some extra safety features.

kayabaNerve committed:
Add support for writing recursive objects.

Also adds checks that the length delimited buffer doesn't exceed its max 
length, and it should do so at every step, stopping any program from 
hitting the callDepthLimit. That said, this is untested.

Related to 
https://github.com/kayabaNerve/nim-protobuf-serialization/issues/12 and 
https://github.com/kayabaNerve/nim-protobuf-serialization/issues/5.

kayabaNerve committed:
Fix https://github.com/kayabaNerve/nim-protobuf-serialization/issues/10.

kayabaNerve committed:
Fix https://github.com/kayabaNerve/nim-protobuf-serialization/issues/11.

kayabaNerve committed:
Use check instead of assert.

kayabaNerve committed:
Fix https://github.com/kayabaNerve/nim-protobuf-serialization/issues/12.

Also updates the length-delimited tests.

kayabaNerve committed:
Extend the bool test.

kayabaNerve committed:
Extend objects test.

Also includes more distinct/enum work. They're grouped with objects as 
they're user defined types. The matching distinct converter exemplifies 
how to use the Protobuf built-in converters with distinct types by 
casting to the underlying type.

kayabaNerve committed:
Fix https://github.com/kayabaNerve/nim-protobuf-serialization/issues/13.

kayabaNerve committed:
Support reading nested objects.

kayabaNerve committed:
Add the out of order test to test_objects.

kayabaNerve committed:
Fix https://github.com/kayabaNerve/nim-protobuf-serialization/issues/1.

Adds extra checks around invalid data.

kayabaNerve committed:
Add a partial object test.

kayabaNerve committed:
Fix https://github.com/kayabaNerve/nim-protobuf-serialization/issues/3.

kayabaNerve committed:
Style changes and bug fixes.

kayabaNerve committed:
More cleanups and bug fixes.

We wrote 11 byts for negative VarInts, not 10.

kayabaNerve committed:
Add mostly function Optional support.

Doesn't test some complex cases and some zigzag VarInts fail to decode. 
I believe this is representative of a wider prolem with VarInts though.

This problem is a blocker for 
https://github.com/kayabaNerve/nim-protobuf-serialization/issues/5 and 
highly eligible for 
https://github.com/kayabaNerve/nim-protobuf-serialization/issues/6.

kayabaNerve committed:
Fix negative PInt reading.

kayabaNerve committed:
Have the failing VarInt tests commented out.

Some of these, like the LibP2P ones, should be commented. Some show 
problems with the underlying code. I do want this independent VarInt 
implementation to work BESIDES the LibP2P tests before moving to Stew's. 
That way, we have a baseline.

kayabaNerve committed:
Clean the code up.

Part of 
https://github.com/kayabaNerve/nim-protobuf-serialization/issues/6.

Greatly removes redundancies/pointless code in the Options macros.

kayabaNerve committed:
Add a proper error for https://github.com/kayabaNerve/nim-protobuf-serialization/issues/14.

Failure tests are added, but since failures don't actually check that 
the error was printed, solely that it failed to compile, these are 
somewhat pointless.

kayabaNerve committed:
More work on Options.

This has truly reached a tipping point for 
https://github.com/kayabaNerve/nim-protobuf-serialization/issues/6. By 
combining that issue with 
https://github.com/kayabaNerve/nim-protobuf-serialization/issues/8, I 
think this library is salavageable. That said, without those, this will 
be impossible to maintain and not submittable.

I want to finish Option support so there's an established baseline. From 
there, I'll do some initial cleanup, and then finally move to the new 
setters and finish cleaning.

kayabaNerve committed:
Bug fixes.

Creates as much of a baseline as possible with the current schema. The 
options test has the recursive type commented because Nim generates 
invalid C code for it.

kayabaNerve committed:
Remove warnings, start cleanup.

kayabaNerve committed:
Start of cleanup.

kayabaNerve committed:
Implement flatMap/box macros.

Replaces createActualTypeFromPotentialOption and getActualType. This 
breaks the library as of this commit, yet is pushed to demonstrate where 
the library is going.

kayabaNerve committed:
Redo the writer.

Still working some bugs out. Part of 
https://github.com/kayabaNerve/nim-protobuf-serialization/issues/6.

kayabaNerve committed:
Fix https://github.com/kayabaNerve/nim-protobuf-serialization/issues/7.

kayabaNerve committed:
Minor cleanup to Writer's hasPragma checks.

kayabaNerve committed:
Fix leftover bugs in the Writer.

It still doesn't detect buffers which have too long of a length, so some 
tests still need work.

kayabaNerve committed:
Replace ProtobufWriter.buffer() with ProtobufWriter.finish().

kayabaNerve committed:
Add basic buffer length checks.

I'm not entirely sure these will work for all cases. test_objects and 
test_length_delimited need expansion.

kayabaNerve committed:
Redo the Reader.

The Writer was verified to work with the old Reader which passed all the 
tests. The new Reader struggles with VarInts and Options. That said, 
these are minor bugs in a set of changes so large I would like to push 
them to clean the diff.

kayabaNerve committed:
Fix reading VarInts.

When working with the type classes, I misdefined the UIntWrappedTypes.

kayabaNerve committed:
Fix #6 and #18.

kayabaNerve committed:
Return None for master objects when there is nothing.

kayabaNerve committed:
Fix https://github.com/kayabaNerve/nim-protobuf-serialization/issues/16.

kayabaNerve committed:
Implement a macro which checks if a type has an overload available.

It's commented out as it's not usable from the library, unfortunately, 
but I hope it'll eventually be usable as it'll be a great improvement.

kayabaNerve committed:
Add support for seqs.

This support is limited; a seq of ints requires being a seq[PInt(int32)] 
or similar. It doesn't map the pragma on the seq. That said, there is a 
function for this in strutils.

kayabaNerve committed:
Add support for arrays.

kayabaNerve committed:
Add support for cstrings.

kayabaNerve committed:
Add the stdlib test I forgot to push.

kayabaNerve committed:
Add support for sets

kayabaNerve committed:
Add support for HashSets.

kayabaNerve committed:
Merge the set/HashSet readers.

kayabaNerve committed:
Don't encode the wireType for every seq item.

Currently still length delimited, even for VarInts/Fixed. That said, it 
properly handles blank elements in both arrays and seqs.

Also fixes the wireType exported template as it didn't properly handle 
numerous cases (refs/ptrs/Options/bools))

kayabaNerve committed:
Commit libp2p's VarInt fully functional and compliant code.

kayabaNerve committed:
Consolidate all existing VarInt code in this library.

Technically not all, as it doesn't include the read/write code. Those 
are too integrated with the Reader/Writer and this is meant to be a 
quick, compliant commit.

kayabaNerve committed:
Move encodeVarInt to varint from writer.

kayabaNerve committed:
Utilize unsafeMemoryInput for the performance advantages.

kayabaNerve committed:
Refactor the API for nim-serialization.

Next is updating the tests to this new API, before moving them over to 
work through nim-serialization entirely.

kayabaNerve committed:
Port over the tests.

Just to sanity check this midstep. The VarInt test is commented. Most of 
these tests have a wrapper function offering the old API, yet some have 
templates to make them cleaner (which also wraps the new API).

kayabaNerve committed:
Delete the libp2p VarInt test.

I want to add back the original VarInt test later.

kayabaNerve committed:
Port every test to encode/decode.

kayabaNerve committed:
Copy over the Serialization generic test.

Unfortunately, due to it assuming integers aren't ambiguous, it's widely 
incompatible. I've added basic pragmas and commented out what doesn't 
work.

https://github.com/kayabaNerve/nim-protobuf-serialization/issues/4 is 
the main blocker.

cheatfate commented:
Maybe i missed something:

1. I don't see usage of protobuf field indexes. Is this indexes are implicit and depending on order in object?
2. I don't see tests for nested objects, which is common for protobuf.

arnetheduck commented on protobuf_serialization/libp2p_varint.nim:
```diff
@@ -0,0 +1,392 @@
+## - LibP2P varint, which is able to encode only 63bits of uint64 number and
+##   maximum size of encoded value is 9 octets (bytes).
+##   https://github.com/multiformats/unsigned-varint
+import bitops, typetraits
```
this needs to be ported to bitops2 - `bitops` has some UB that we don't really want to be burdened by

arnetheduck commented:
> I don't see usage of protobuf field indexes.

+1 - field indices are a key aspect of protocol definitions in protobuf - they should not be optional / derived / implicit in the object definitions - they form the keystone for protocol forward/backward compatibility - it's important to stress that protobuf first and foremost is an interchange format between processes and being explicit is a feature here, as one of the design goals of it is to maintain compatibility in a dynamic world

zah commented on protobuf_serialization/internal.nim:
```diff
@@ -0,0 +1,156 @@
+  #While cstring/array are built-ins, and therefore should have converters provided, but they still need converters.
+  LengthDelimitedTypes* = not (VarIntTypes or FixedTypes)
+
+proc getTypeChain(impure: NimNode): seq[NimNode] {.compileTime.} =
```
Some documentation here would be appropriate.

As far as I can tell, it's quite possible that you don't need this at all, because you can implement `flatType` as something close to `type(flatMapInternal(value))`.

`isPotentiallyNull` will be a simple recursive function replicating the logic in `getTypeChain`, but without the error-prone type AST business.

zah commented on protobuf_serialization/internal.nim:
```diff
@@ -0,0 +1,156 @@
+  quote do:
+    `variable` = `wrap`
+
+macro isStdlib*(ty: untyped): untyped =
```
This is definetely not needed. What's wrong with using the `is` operator directly everywhere where `isStdLib` is used and having simpler checks such as:

```nim
when value is seq:
   writeSeq(value)
elif value is set:
   writeSet(value)
...
```

zah commented on protobuf_serialization/writer.nim:
```diff
@@ -0,0 +1,251 @@
+  if false:
+    raise newException(IOError, "")
+
+  let flattenedOption = value.flatMap()
```
The `flatMap` approach leads to unnecessary copies though.

With the recursive approach, it would be very easy to traverse the Options and refs exactly in the same way `flaptMap` currently traverses them. In the end, the recursive process will have two possible outcomes:

1) It will terminate, because it found a `nil`/`none` value (this is similar to the `return` statement below)

2) It will reach a leaf value and the serialization will continue. The leaf value will be equivalent to the `flattened` variable below, so you can just copy paste the entire block of code and place it in the function that will be eventually called by the recursive process.

zah commented on protobuf_serialization/reader.nim:
```diff
@@ -0,0 +1,295 @@
+      preResult = T(value)
+
+  #Box the result back up.
+  box(fieldVar, preResult)
```
If you use a reader table, like in the Json library, the copy that `box` introduces here can be avoided.

The logic will become the following:

1) Read a field ID from the stream
2) Check if we have a reader associated with the ID (the ID may be bogus)
3) If we have a reader, call it by passing the stream

Now, the reader knows the type it wants to construct and it already knows that the field is present, so it can unconditionally construct the "box" in which the final value will be placed (either a ref allocation or a `Option[T]` value). Once the box is constructed, it just calls the normal `readValue` function for the type to read the actual value from the stream.

zah commented on protobuf_serialization/reader.nim:
```diff
@@ -0,0 +1,295 @@
+
+  var
+    bytes = newSeq[byte](stream.eofSafeRead())
+    preResult: B
```
After reading the length of the sequence here, you should check if the stream can buffer enough bytes with `stream.readable(n)`.

`n` should be capped to some maximum value, otherwise a remote attacker can force the code to allocate too much memory and crash the process.

Once the stream has buffered enough bytes (`stream.readable(n)` returns true), you can use the `stream.withReadableRange` API to temporary restrict the readable portion of the stream before you pass it to `stdlibFromProtobuf`.

`stdlibFromProtobuf` should be actually broken into separate functions responsible for the different std lib types as indicated in another comment. Each such function should take an `InputStream` instead of `seq[byte]` and it will consume the data naturally with all the other existing `readValue` functions that work with streams already. No need for byte sequence allocations and `unsafeMemoryInput` as done below.

zah commented on protobuf_serialization/libp2p_varint.nim:
```diff
@@ -0,0 +1,392 @@
+                   typetraits.name(type(value)) & "]")
+
+proc encodeVarint*(vtype: typedesc[PB],
+                   value: PBSomeVarint): seq[byte] {.inline.} =
```
This can write directly to an output stream. This way, it can be asked to populate a regular array on the stack with `unsafeMemoryOutput`, or it can write to an arbitrary output stream without allocating a small sequence and copying its bytes over.

zah commented on protobuf_serialization/writer.nim:
```diff
@@ -0,0 +1,251 @@
+  if sub:
+    existingLength += 2
+
+  var bytes: seq[byte]
```
No need for buffering here. You can use the ["delayed writes"](https://github.com/status-im/nim-faststreams/#delayed-writes) feature of the output stream to write the varint value after all the other bytes are written.

There will be no temporary `memoryOutputs` here and all output functions will directly work with an `OutputStream` instead of populating a sequence first.

kayabaNerve commented:
Thanks for all the feedback.

Yes, field numbering is implicit. That said, as only serialized field numbers get numbered, it's fully stable even when modifying unserialized fields. I'm happy to make it explicit.

tests/test_objects does test nested objects.

libp2p_varint isn't used at all. I have yet to consolidate libp2p_varint with the custom varint implementation in this codebase. When I do that, I'll be sure to use bitops2. https://github.com/kayabaNerve/nim-protobuf-serialization/issues/9 sort-of tracks this. It hasn't been commented to be up to date with the current codebase layout, yet covers the idea.

The only AST functions I'm explicitly happy with are the flatten ones. That said, flatMapInternal does require typechain. flatMapInternal always returns the base value. When any step is None/nil, it returns none(base). That said, I will say I'm not sure we ever take advantage of this property. It may have been done just for a constant return type signature, but that may be work-aroundable with auto, and definitely is with untyped.

The reason toStdlib exists is to remove code duplication. I can get rid of it if preferred (which it sounds like it is), which also allows getting rid of the overloaded standard lib writers.

Implementation of a reader table is an open issue under https://github.com/kayabaNerve/nim-protobuf-serialization/issues/8.

Thanks for the tips about removing seq[byte], zah. I do have these inefficiencies noted in https://github.com/kayabaNerve/nim-protobuf-serialization/issues/20. I appreciate the heads up on API tricks to efficiently remove those.

I, personally, wouldn't consider this library 'complete' due to:

- https://github.com/kayabaNerve/nim-protobuf-serialization/issues/20 (seq[byte])
- https://github.com/kayabaNerve/nim-protobuf-serialization/issues/9 (multiple VarInt implementations and I'm not convinced mine is compliant)
- https://github.com/kayabaNerve/nim-protobuf-serialization/issues/4 (shoddy standard lib type support).

That said, I appreciate all the feedback as I work more on this.

zah commented:
The ProtoBuf docs have the following to say about the field numbers:
> As you can see, each field in the message definition has a unique number. These field numbers are used to identify your fields in the message binary format, and should not be changed once your message type is in use. Note that field numbers in the range 1 through 15 take one byte to encode, including the field number and the field's type (you can find out more about this in Protocol Buffer Encoding). Field numbers in the range 16 through 2047 take two bytes. So you should reserve the numbers 1 through 15 for very frequently occurring message elements. Remember to leave some room for frequently occurring elements that might be added in the future. 

Since it's recommend to reserve field numbers for future use, it's quite likely that the implicit numbering will be off quite often and it seems better to force the user to think about this problem. Also, I guess a library should follow the spirit and the design choices of the standard being implemented, so let's make the field numbers a mandatory pragma.

zah commented:
> The reason toStdlib exists is to remove code duplication. I can get rid of it if preferred (which it sounds like it is), which also allows getting rid of the overloaded standard lib writers.

To clarify my previous comment, writing correct type matching code on the AST level (as the current code tries) is an extremely difficult task because the type system of Nim has very complicated rules for type equivalence (you have to take care of type aliases, generic type aliases, resolved generic parameter types, instantiated concept types and so on).

You should try very hard to avoid doing this sort of thing in macros - it's much easier to write correct code by relying on the `is` operator in regular templates and generic functions. When you need to compute a certain type depending on input expressions, think about whether you can employ regular overloads and the `type()` operator to your advantage. For example, here is how I would have defined `flatType`:

```nim
template deref(x: auto): auto =
  when x is Option:
    deref(x.get)
  elif x is ref:
    deref(x[])
  else:
    x

template flatType(x: auto): type =
  type(deref(x))

var x: ref Option[ref int]

type T = flatType(x)
static: assert T is int
```

With that said, I still think you should go for the recursive solution in order to avoid the `flattened` copy.

kayabaNerve committed:
Move reading entirely to streams.

toProtobuf/fromProtobuf are still streamless, as there's no point in 
moving them over when we're getting rid of them.

Doesn't work with faststreams due to 
https://github.com/status-im/nim-faststreams/issues/12.

Also removes string from a castable type due to problems with \0.

kayabaNerve committed:
Finally consolidate the readVarInt with the VarInt file.

kayabaNerve committed:
Fix https://github.com/kayabaNerve/nim-protobuf-serialization/issues/23.

kayabaNerve committed:
Fix seq[object] encoding/decoding.

Part of 
https://github.com/kayabaNerve/nim-protobuf-serialization/issues/4.

kayabaNerve committed:
WIP. Remove array overhead and use VarInts for buffer lengths.

Makes progress on 
https://github.com/kayabaNerve/nim-protobuf-serialization/issues/21 and 
https://github.com/kayabaNerve/nim-protobuf-serialization/issues/4.

Breaks some existing tests and doesn't test the new functionality. 
Committing because I'm set on using this code and it's already a large 
commit. Also have some misc tasks to care of before I finish this.

kayabaNerve committed:
Remove the raises pragma.

As we raised Defect, and not all Exceptions were raised in all 
circumstances, it wasn't too helpful and caused `if false: raise X`.

kayabaNerve committed:
Fix the not-working seq code.

Releated to the WIP commit attached to 
https://github.com/kayabaNerve/nim-protobuf-serialization/issues/4.

kayabaNerve commented:
I would like to note all checks do pass locally, with https://github.com/status-im/nim-faststreams/issues/12 being the blocker for the CI.

kayabaNerve committed:
Theoretically working streamed writer.

kayabaNerve committed:
isPotentiallyNull/flatType cleanup.

kayabaNerve committed:
Redo flatMap.

kayabaNerve committed:
Add a prototype box redo, commented out.

kayabaNerve committed:
Rewrite box using is.

kayabaNerve committed:
Remove unused macro imports.

kayabaNerve committed:
Support pointers.

kayabaNerve committed:
Cause intended behavior of https://github.com/kayabaNerve/nim-protobuf-serialization/issues/16.

kayabaNerve committed:
Pure toProtobuf/fromProtobuf.

Part of 
https://github.com/kayabaNerve/nim-protobuf-serialization/issues/20.

kayabaNerve committed:
Fix https://github.com/kayabaNerve/nim-protobuf-serialization/issues/21.

kayabaNerve committed:
Cleanup.

Turns verifyWritable into verifySerializable, now called from the Reader 
as well.

Removes export symbols from functions which should not be exported.

Remove eofSafeRead now that it only has one caller.

Deletes unused functions.

kayabaNerve committed:
Add support for tuples.

kayabaNerve committed:
Track the root type in stdlib To/From Protobuf.

Necessary to get its pragmas information.

kayabaNerve committed:
Enable attaching pragmas to stdlib types.

The test fails, yet for unrelated reasons. array[bool] and 
HashSet[float] have some problems I'm still debugging.

Enables removing required from 
https://github.com/kayabaNerve/nim-protobuf-serialization/issues/4.

kayabaNerve committed:
Add back a required clause.

kayabaNerve committed:
Fix stdlib[boolean] encoding.

kayabaNerve committed:
Fix stdlib[Fixed] decoding.

kayabaNerve committed:
Merge PInt and UInt. Add LibP2P Int pragmas/

PInt and UInt follow the same encoding rules for positive values, except 
the sign bit in PInt is used for the value bit in UInt. There's also no 
way this can be ambiguous.

I do worry about the fact it may encourage PInt usage, when SInt should 
be used for signed values (when deciding which type to use, not when 
matching compatibility). That said, it's incorrect to group UInt with 
SInt due to vastly different encoding methods.

kayabaNerve committed:
Rework VarInt type classes.

kayabaNerve committed:
Implement viSizeof and a faster encode.

kayabaNerve committed:
Consolidate encoding API with libp2p_varint.

Includes further optimizations.

kayabaNerve committed:
Move a runtime check to compile time.

Probably already optimized out.

kayabaNerve committed:
Removed signed LibP2P int.

LibP2P only handles unsigned VarInts.

kayabaNerve committed:
Rewrite decode.

Seperates VarInt extracting from decoding.

kayabaNerve committed:
Finish merging libp2p_varint with varint.

There's still some API differences I want to resolve, not to mention the 
move from types to enums.

I also want to clean the type classes up a bit.

Part of 
https://github.com/kayabaNerve/nim-protobuf-serialization/issues/9.

kayabaNerve committed:
Fix https://github.com/kayabaNerve/nim-protobuf-serialization/issues/15.

kayabaNerve committed:
Require explicit specification of field numbers.

Part of 
https://github.com/kayabaNerve/nim-protobuf-serialization/issues/22.

kayabaNerve commented:
Once I add a few minor enhancements which are bugging me (related to field numbers; we currently allow duplicates/gaps), the only blockers before merging are:

- Lack of more comprehensive tests. I have plenty, but they don't check against any other implementation, or some failure cases.
- Lack of a streamed writer. I have a prototype on a branch but it isn't merged.

The VarInt code is marked as a requirement before merging (API/submission back to Stew/Fastreams/its own lib), but as long as it's valid in implementation, that isn't a requirement to complete before merging.

That still leaves more enhancements//optimizations, of course, yet nothing worth delaying the merging of this library.

kayabaNerve commented:
Reading the above commentary again to confirm I resolved all issues (or am explicitly tracking the rest ), which I believe I'm doing, I did realize a protocol compliance issue I have to fix related to field numbers. I've correctly marked it accordingly.

kayabaNerve committed:
Merge branch 'streamed-writer'.

Mostly functional. One edge case still fails due to a merge conflict I'm 
still tracking down.

kayabaNerve committed:
Fix a failing test related to the streamed writer and a couple warnings.

kayabaNerve committed:
Don't allow field number reuse.

Part of 
https://github.com/kayabaNerve/nim-protobuf-serialization/issues/22.

kayabaNerve committed:
Allow gap fields.

Part of 
https://github.com/kayabaNerve/nim-protobuf-serialization/issues/22.

kayabaNerve commented:
It's solely down to tests now. While it isn't ready for integration into LibP2P, it is viable for merging if desired.

kayabaNerve committed:
Add a missing close comment I forgot to commit.

kayabaNerve committed:
Add failure tests to verify bit/encoding/field number requirements.

kayabaNerve committed:
Correct some type classes.

kayabaNerve committed:
Correct the supported types check using a fake write.

Explicitly bans tuples and tables.

kayabaNerve committed:
Test string encoding preserves the null terminator.

kayabaNerve committed:
Correct a check in VarInt.

kayabaNerve committed:
Add a fixed test.

Also fixes a commit bug in VarInt and uses void for types which should 
never be utilized.

kayabaNerve committed:
Remove a flag from the Nimble file which isn't needed.

kayabaNerve committed:
Correct the fixed test.

kayabaNerve committed:
Consolidate Fixed code. Cleaning.

Don't both wrapping types which only have one valid wrapped 
representation (marks byte/bool/char/float32/float64 as wrapped 
already).

Consolidates type classes and updates choice of type classes to remove 
uneeded cases.

Adds missing lint/LInt case handlers.

Also causes 
https://github.com/kayabaNerve/nim-protobuf-serialization/issues/26 to 
exist. I have no idea how it did work previously.

kayabaNerve committed:
Fix https://github.com/kayabaNerve/nim-protobuf-serialization/issues/26.

kayabaNerve committed:
Check that length delimited buffers use a proper VarInt length encoding.

kayabaNerve committed:
Finish support for LInt.

Adds a proper test.

zah commented on protobuf_serialization.nim:
```diff
@@ -1,376 +1,18 @@
+                    Writer = ProtobufWriter,
+                    PreferedOutput = seq[byte]
+
+template supports*(_: type Protobuf, T: type): bool =
```
`supports` is meant to be something like a compile-time check that a particular type can be serialized.
It shouldn't emit any run-time code.

But since this is not a critical feature, it can be addressed after the initial merge

kayabaNerve replied:
Fix in https://github.com/kayabaNerve/nim-protobuf-serialization/commit/70a7274f0a62d892f1c8081b1dfd6c835d5ca64f. That said, it doesn't return a bool, instead using fatal.

zah commented on protobuf_serialization/reader.nim:
```diff
@@ -0,0 +1,193 @@
+            hasPInt = T.hasCustomPragmaFixed(fieldName, pint)
+            hasSInt = T.hasCustomPragmaFixed(fieldName, sint)
+            hasLInt = T.hasCustomPragmaFixed(fieldName, lint)
+            hasFixed = T.hasCustomPragmaFixed(fieldName, fixed)
```
We can add a sanity check that there are no conflicting pragmas by summing these constants.

kayabaNerve replied:
Already done in verifySerializable, which every written/read object goes through, even when nested and the nested object is only written through the parent. I'll add a failure test for this specific case.

kayabaNerve replied:
Confirmed in https://github.com/kayabaNerve/nim-protobuf-serialization/commit/f492d9470cc0c00140aa1973a0501ba19e6c4a17.

zah commented on protobuf_serialization/reader.nim:
```diff
@@ -0,0 +1,193 @@
+      {.fatal: "Tried to read a Length Delimited value which we didn't recognize. This should never happen.".}
+    elif type(preResult) is CastableLengthDelimitedTypes:
+      var byteResult: seq[byte] = newSeq[byte](len)
+      if not substream.readInto(byteResult):
```
The main idea of `readInto` is that it allows you to avoid the intermediate buffering performed by FastStreams if your goal is to store the raw bytes in a long-lived location. Here, you can move it all the way to line 81 and use it instead of `stream.readable`. Just like `stream.readable`, it will return `false` if not enough bytes are present in the stream.

kayabaNerve replied:
Fixed in https://github.com/kayabaNerve/nim-protobuf-serialization/commit/c447e58420935e8dd6812d2a77275a70c848b701.

zah commented on protobuf_serialization/stdlib_readers.nim:
```diff
@@ -0,0 +1,186 @@
+      if len < 0:
+        raise newException(ProtobufMessageError, "String longer than 2 GB specified.")
+
+      stream.withReadableRange(len, substream):
```
You need a add a check `if stream.readable(len)` before entering the readable range here.
I guess it would be nice to add a new helper is FastStreams that performs the two steps at once.

kayabaNerve replied:
Fixed in https://github.com/kayabaNerve/nim-protobuf-serialization/commit/95608cc3c67737100bd17675be9aca36a4c9d4db.

zah commented on protobuf_serialization/numbers/varint.nim:
```diff
@@ -0,0 +1,268 @@
+    if inLen == bytes.len:
+      return VarIntStatus.Incomplete
+    next = bytes[inLen]
+    value += (next and U(VAR_INT_VALUE_MASK)) shl offset
```
What would prevent a possible int overflow here?

kayabaNerve replied:
Nothing. Just added one, will push when tested.

kayabaNerve replied:
Fixed in https://github.com/kayabaNerve/nim-protobuf-serialization/commit/9b8654fd02346b59f473ce8d6f43646f54e8efc7.

kayabaNerve replied:
Also https://github.com/kayabaNerve/nim-protobuf-serialization/commit/bda52c8591cffeb0f8cb61c369d31fb16f123e42.

zah commented on protobuf_serialization/numbers/fixed.nim:
```diff
@@ -0,0 +1,84 @@
+  else:
+    var temp: uint32
+  for i in 0 ..< sizeof(temp):
+    temp = temp + (type(temp)(stream.read()) shl (i * 8))
```
You have to check that the stream is readable here.

kayabaNerve replied:
Thanks. Fixed.

zah commented on protobuf_serialization/stdlib_readers.nim:
```diff
@@ -0,0 +1,186 @@
+  stream: InputStream,
+  _: typedesc[R],
+  unusedFieldName: static string,
+  value: var cstring
```
This doesn't look safe. The `cstring` is set to point to a orphaned Nim string, which the GC can collect at some point. I think you should remove the support for `cstring` altogether.

kayabaNerve replied:
Disabled cstrings in https://github.com/kayabaNerve/nim-protobuf-serialization/commit/e18579a4336c0787fd3478ac534888139f5915da. I did have a test for this, but I didn't implement a GC_fullCollect in said test.

zah commented on protobuf_serialization/stdlib_readers.nim:
```diff
@@ -0,0 +1,186 @@
+        raise newException(ProtobufEOFError, "Length delimited buffer doesn't have enough data to read the next object.")
+
+      stream.withReadableRange(len, substream):
+        ProtobufReader.init(substream, closeAfter = false).readValue(seqInstance[^1])
```
Why is the fresh `ProtoBufReader` needed here?

Perhaps only because this function is not taking a reader?

kayabaNerve replied:
I'd have to overwrite the Reader's stream + properties with substream/false and set it back. I didn't bother to add the Reader to every stdlib_reader for that small optimization.

zah commented on protobuf_serialization/stdlib_readers.nim:
```diff
@@ -0,0 +1,186 @@
+  stream: InputStream,
+  ty: typedesc[R],
+  fieldName: static string,
+  seqInstance: var seq[T]
```
I think arrays and sequences should have been treated with the semantics of the ProtoBuf `repeated` keyword.
That is, you keep going through the buffer and you keep finding elements with the same field number. The elements doesn't even have to appear consequently as far as ProtoBuf is concerned. Every time you find such an element, you append it to the array/sequence.

kayabaNerve replied:
Yep. This is being marked as required. My bad.

kayabaNerve replied:
This is... complicated. https://github.com/kayabaNerve/nim-protobuf-serialization/issues/29 covers the problem.

zah commented on protobuf_serialization/writer.nim:
```diff
@@ -0,0 +1,159 @@
+  if (stream.pos != startPos) or (rootType.isPotentiallyNull()):
+    cursor.finalWrite(newProtobufKey(fieldNum, LengthDelimited) & encodeVarInt(PInt(int32(stream.pos - startPos))))
+  else:
+    cursor.finalWrite([])
```
This code will write zero bytes in place of the length prefix (effectively omitting the prefix). I suspect this is not what you wanted to do. You probably wanted to write a single zero byte?

kayabaNerve replied:
If the length is 0, we don't encode the buffer at all. It's omitting it. Since we encode the field key with the length, when there is one, this is valid.

zah commented on protobuf_serialization/numbers/fixed.nim:
```diff
@@ -0,0 +1,84 @@
+    var casted = cast[uint32](value)
+
+  for _ in 0 ..< sizeof(casted):
+    stream.write(byte(casted and LAST_BYTE))
```
We have some helpers in `stew/endians2` that produce BigEndian and LittleEndian byte sequences.

kayabaNerve replied:
Opened an issue for it.

zah commented on protobuf_serialization/writer.nim:
```diff
@@ -0,0 +1,159 @@
+  elif flattened is (object or tuple):
+    enumInstanceSerializedFields(flattened, fieldName, fieldVal):
+      discard fieldName
+      const fieldNum = getCustomPragmaVal(fieldVal, fieldNumber)
```
We should try to provide a helpful error message if the user has failed to add the `fieldNumber` pragma

kayabaNerve replied:
A helpful error message requires getCustomPragmaFixed which I couldn't get working due to symbol resolution errors. I'll open an issue for it.

kayabaNerve committed:
Make supports compile time.

kayabaNerve committed:
Add a test to verify nested objects get verified as serializable.

kayabaNerve committed:
Optimize CLDT.

kayabaNerve committed:
Add a missing readable check.

kayabaNerve committed:
Fix VarInt overflow checks.

kayabaNerve committed:
And low(X) checks. Fix related bugs.

kayabaNerve committed:
Add a missing call to readable.

kayabaNerve committed:
Disable cstrings.

Preserves the decoder in case we ever want to re-enable them.

kayabaNerve committed:
Add the always unpacked writer from repeated.

kayabaNerve committed:
Only use unpacked buffers when packed buffers can't be used.

kayabaNerve committed:
Working packIntoSeq.

kayabaNerve committed:
Misc bug fixes.

kayabaNerve committed:
Support object fields which aren't sequences.

kayabaNerve committed:
Bug fix/remove pointless argument.

kayabaNerve committed:
Bug fixes which get most tests passing with this branch.

kayabaNerve committed:
Confirm every key is valid.

The unpacked -> packed writer stripped invalid keys (and only the key; 
not any following data).

kayabaNerve committed:
Correct the check to prefix the length.

kayabaNerve committed:
Correct input stream closing.

kayabaNerve committed:
Correct placement of the stream.readable check in reader.

kayabaNerve committed:
Use a try/finally to verify input streams aren't leaked.

kayabaNerve committed:
Isolate the stream packing code into its own file.

Also merges code used in both the reader/writer to internal.

kayabaNerve committed:
Removing packing and disable arrays.

kayabaNerve committed:
Make progress on supporting unpacked seqs.

kayabaNerve committed:
Restore the field number existence check.

kayabaNerve committed:
Fix seq decoding.

kayabaNerve committed:
Fix https://github.com/kayabaNerve/nim-protobuf-serialization/issues/29.

kayabaNerve committed:
Merge pull request #31 from kayabaNerve/native-unwrapped

Fix https://github.com/kayabaNerve/nim-protobuf-serialization/issues/29.

kayabaNerve committed:
Correct faulty tests.

kayabaNerve committed:
Fix the Nimble package file.

kayabaNerve committed:
Support enums without negative values with unsigned encodings.

kayabaNerve committed:
Add a missing flatten.

kayabaNerve committed:
Add a dontOmit pragma which matches Protobuf's required.

kayabaNerve committed:
Add VarInt length prefixing.

kayabaNerve committed:
Add a missing cursor close.

kayabaNerve committed:
Fix seq[seq[byte]] encoding.

kayabaNerve committed:
Remove LInt except via encodeVarInt.

Removes ProtobufWriteError.

kayabaNerve commented:
Beyond the ongoing discussion around how to resolve Proto2/Proto3 differences, this is ready to be merged IMO. Depending on how those discussions go, https://github.com/kayabaNerve/nim-protobuf-serialization/issues/32 may be marked a requirement before merging, or the addition of proto2/proto3 pragmas may be required.

kayabaNerve committed:
Minor cleanup.

Fixes 
https://github.com/kayabaNerve/nim-protobuf-serialization/issues/30.

zah commented:
Merging to make this more easily accessible for the Waku team. Further reviews can be added here and addressed later.
